vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,495 @@
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/vm/proxyapp/proxyrpc"
+	"github.com/google/syzkaller/vm/vmimpl"
+)
+
+func osutilCommand(bin string, args ...string) subProcessCmd {
+	return osutil.Command(bin, args...)
+}
+
+type subProcessCmd interface {
+	StdinPipe() (io.WriteCloser, error)
+	StdoutPipe() (io.ReadCloser, error)
+	StderrPipe() (io.ReadCloser, error)
+	Start() error
+	Wait() error
+}
+
+func makeDefaultParams() *proxyAppParams {
+	return &proxyAppParams{
+		CommandRunner:   osutilCommand,
+		MakePipe:        osutil.LongPipe,
+		InitRetryDelay:  10 * time.Second,
+		MaxInitAttempts: math.MaxInt64,
+	}
+}
+
+func init() {
+	vmimpl.Register(
+		"proxyapp",
+		func(env *vmimpl.Env) (vmimpl.Pool, error) {
+			return ctor(makeDefaultParams(), env)
+		},
+		false)
+}
+
+type pool struct {
+	mu              sync.Mutex
+	isValid         bool
+	env             *vmimpl.Env
+	Proxy           *ProxyApp
+	ctxRunner       context.Context
+	ctxRunnerCancel context.CancelFunc
+	size            int
+}
+
+func (pool *pool) reset() {
+	pool.cancelRun()
+	pool.ctxRunner, pool.ctxRunnerCancel = context.WithCancel(context.TODO())
+	pool.Proxy = nil
+	pool.size = 0
+	pool.isValid = false
+}
+
+func (pool *pool) cancelRun() {
+	if pool.ctxRunnerCancel != nil {
+		pool.ctxRunnerCancel()
+		pool.ctxRunnerCancel = nil
+	}
+}
+
+func initPool(params *proxyAppParams, pool *pool, command string, pluginConfig []byte, env *vmimpl.Env,
+) error {
+	pool.reset()
+
+	var err error
+	pool.Proxy, err = runProxyApp(pool.ctxRunner, params, command)
+	if err != nil {
+		return fmt.Errorf("failed to run ProxyApp: %w", err)
+	}
+
+	pool.size, err = pool.Proxy.ConstructPool(string(pluginConfig), env.Debug)
+	if err != nil {
+		pool.cancelRun()
+		return fmt.Errorf("failed to construct pool: %w", err)
+	}
+
+	pool.isValid = true
+	return nil
+}
+
+type Config struct {
+	Command        string          `json:"proxy_cmd"`
+	ProxyAppConfig json.RawMessage `json:"proxyapp_config"`
+}
+
+func parseConfig(conf []byte) (*Config, error) {
+	vmCfg := new(Config)
+	if err := config.LoadData(conf, vmCfg); err != nil {
+		return nil, fmt.Errorf("failed to parseConfig(): %w", err)
+	}
+	return vmCfg, nil
+}
+
+// Package configuration VARs are mostly needed for tests.
+type proxyAppParams struct {
+	CommandRunner   func(bin string, args ...string) subProcessCmd
+	MakePipe        func() (io.ReadCloser, io.WriteCloser, error)
+	InitRetryDelay  time.Duration
+	MaxInitAttempts int64
+}
+
+func ctor(params *proxyAppParams, env *vmimpl.Env) (vmimpl.Pool, error) {
+	subConfig, err := parseConfig(env.Config)
+	if err != nil {
+		return nil, fmt.Errorf("config parse error: %w", err)
+	}
+
+	pool := &pool{
+		env: env,
+	}
+	pool.mu.Lock()
+	minWait := time.After(0)
+	initialized := make(chan bool)
+	attempt := int64(0)
+
+	go func() {
+		for {
+			attempt++
+			if attempt > params.MaxInitAttempts {
+				pool.mu.Unlock()
+				initialized <- false
+				return
```

Comment:
What is supposed to happen next? Will a syz-manager instance be able to gracefully handle a never-from-this-point-working vm pool?

Reply:
No. I expect syz-manager to stop trying here.
Low MaxInitAttempts are used for testing.
High MaxInitAttemps (math.MaxInt64) will be used for deployment.

Reply:
I mean, what will syz-manager itself do?
From the `manager.go` code it looks as if it will just indefinitely loop in `try to create a VM -> fail -> try again`. Is it what we want? Maybe we need some rate limiting (not to garbage the logs) and some error threshold before syz-manager process stops trying and exits.

Reply:
syz-manager will never go initialized := false path.
And in the updated code you can see that pool.Create() blocks and wait for the initialized pool.proxy instance.
Additionally, pool initialization attempts are rate limited by InitRetryDelay:  10 * time.Second.